package/kedro_viz/api/rest/responses/pipelines.py

```diff
@@ -0,0 +1,208 @@
+
+    id: str
+    type: str
+
+
+class ModularPipelinesTreeNodeAPIResponse(BaseAPIResponse):
+    """Model a node in the tree representation of modular pipelines in the API response."""
+
+    id: str
+    name: str
+    inputs: List[str]
+    outputs: List[str]
+    children: List[ModularPipelineChildAPIResponse]
+
+
+# Represent the modular pipelines in the API response as a tree.
+# The root node is always designated with the __root__ key.
+# Example:
+# {
+#     "__root__": {
+#            "id": "__root__",
+#            "name": "Root",
+#            "inputs": [],
+#            "outputs": [],
+#            "children": [
+#                {"id": "d577578a", "type": "parameters"},
+#                {"id": "data_science", "type": "modularPipeline"},
+#                {"id": "f1f1425b", "type": "parameters"},
+#                {"id": "data_engineering", "type": "modularPipeline"},
+#            ],
+#        },
+#        "data_engineering": {
+#            "id": "data_engineering",
+#            "name": "Data Engineering",
+#            "inputs": ["d577578a"],
+#            "outputs": [],
+#            "children": [],
+#        },
+#        "data_science": {
+#            "id": "data_science",
+#            "name": "Data Science",
+#            "inputs": ["f1f1425b"],
+#            "outputs": [],
+#            "children": [],
+#        },
+#    }
+# }
+ModularPipelinesTreeAPIResponse = Dict[str, ModularPipelinesTreeNodeAPIResponse]
+
+
+class GraphAPIResponse(BaseAPIResponse):
+    nodes: List[NodeAPIResponse]
+    edges: List[GraphEdgeAPIResponse]
+    layers: List[str]
+    tags: List[NamedEntityAPIResponse]
+    pipelines: List[NamedEntityAPIResponse]
+    modular_pipelines: ModularPipelinesTreeAPIResponse
+    selected_pipeline: str
+
+
+def get_pipeline_response(
+    pipeline_id: Union[str, None] = None
+) -> Union[GraphAPIResponse, JSONResponse]:
+    """API response for `/api/pipelines/pipeline_id`."""
+    if pipeline_id is None:
+        pipeline_id = data_access_manager.get_default_selected_pipeline().id
+
+    if not data_access_manager.registered_pipelines.has_pipeline(pipeline_id):
+        return JSONResponse(status_code=404, content={"message": "Invalid pipeline ID"})
+
+    modular_pipelines_tree = (
+        data_access_manager.create_modular_pipelines_tree_for_registered_pipeline(
+            pipeline_id
+        )
+    )
+
+    return GraphAPIResponse(
+        nodes=data_access_manager.get_nodes_for_registered_pipeline(pipeline_id),
+        edges=data_access_manager.get_edges_for_registered_pipeline(pipeline_id),
+        tags=data_access_manager.tags.as_list(),
+        layers=data_access_manager.get_sorted_layers_for_registered_pipeline(
+            pipeline_id
+        ),
+        pipelines=data_access_manager.registered_pipelines.as_list(),
+        modular_pipelines=modular_pipelines_tree,
+        selected_pipeline=pipeline_id,
+    )
+
+
+def get_kedro_project_json_data():
```

Comment:
was wondering if this can be moved to utils.py

Reply:
Hi @rashidakanchwala, from the function name it seems to be a utility, but it is used to provide a pipeline response. Also we use `get_encoded_response` from utils. Placing this in utils will create a circular dependency.